[WPE][GStreamer] support auto-pluging of different sinks

emutavchi opened the pull request:
#### bfa95384feb8c8b235430a59de9e6b9ebff333c9
<pre>
[WPE][GStreamer] support auto-pluging of different sinks
<a href="https://bugs.webkit.org/show_bug.cgi?id=256482">https://bugs.webkit.org/show_bug.cgi?id=256482</a>

Reviewed by Xabier Rodriguez-Calvar.

This solves several cases:

- secondary (video only/pip) playback: audio sink allocates resources in
ready state (as part of auto audio sink detection), which interferes
with the main playback.

- &quot;external&quot; playback, or playback with other sink elements: in this
case we want playbin2 to autoplug needed elements.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::createAudioSink):
(WebCore::MediaPlayerPrivateGStreamer::gstreamerPositionFromSinks const):
(WebCore::MediaPlayerPrivateGStreamer::handleMessage):
(WebCore::MediaPlayerPrivateGStreamer::createHolePunchVideoSink):

Canonical link: <a href="https://commits.webkit.org/263972@main">https://commits.webkit.org/263972@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b4bf638e839558cb17f8fa639272a39601100895

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6236 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6428 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6607 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7804 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6567 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6235 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6648 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6378 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9446 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6346 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/6361 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5646 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7875 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3834 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5625 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13518 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5699 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5705 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/7953 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6183 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5056 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5596 "Built successfully") | | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1483 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9743 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5963 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -1837,6 +1843,21 @@ void MediaPlayerPrivateGStreamer::handleMessage(GstMessage* message)
+        if (currentState <= GST_STATE_READY && newState >= GST_STATE_READY) {
+            // If we didn't create an audio sink, store a reference to the created one.
+            if (!m_audioSink) {
+                // Detect a audio sink element
```
Comment needs to end with a .

calvaris reviewed (changes requested):
Title of the PR has to be in the first line of the commit log. Bug URL has to be in the second line of the log. After these two lines you need to have `Reviewed by NOBODY (OOPS!).`, then the bug explanation (remove the second OOPS line as you already have the explanation.

Please, check some other commits to have an idea of how to do it.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -1837,6 +1843,21 @@ void MediaPlayerPrivateGStreamer::handleMessage(GstMessage* message)
+        if (currentState <= GST_STATE_READY && newState >= GST_STATE_READY) {
+            // If we didn't create an audio sink, store a reference to the created one.
+            if (!m_audioSink) {
+                // Detect a audio sink element.
```
Sorry to not have noticed this before because I asked you already once to change this line:

Detect "an" audio sink element.

emutavchi replied:
Done

calvaris reviewed (approved):
One you push the list last change, I'll already ask the merge queue to merge it. Thanks.

emutavchi committed:
[WPE][GStreamer] support auto-pluging of different sinks
https://bugs.webkit.org/show_bug.cgi?id=256482

Reviewed by Xabier Rodriguez-Calvar.

This solves several cases:

- secondary (video only/pip) playback: audio sink allocates resources in
ready state (as part of auto audio sink detection), which interferes
with the main playback.

- "external" playback, or playback with other sink elements: in this
case we want playbin2 to autoplug needed elements.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::createAudioSink):
(WebCore::MediaPlayerPrivateGStreamer::gstreamerPositionFromSinks const):
(WebCore::MediaPlayerPrivateGStreamer::handleMessage):
(WebCore::MediaPlayerPrivateGStreamer::createHolePunchVideoSink):

Canonical link: https://commits.webkit.org/263972@main
